validator/src/micro.rs

```diff
@@ -84,75 +84,103 @@ impl<TValidatorNetwork: ValidatorNetwork + 'static> NextProduceMicroBlockEvent<T
+            expected_next_ts = self.expected_next_timestamp(&blockchain);
 
-                        let block1 = block.clone();
-
-                        // Use a trusted push since these blocks were generated by this validator
-                        let result = if cfg!(feature = "trusted_push") {
-                            Blockchain::trusted_push(blockchain, Block::Micro(block))
-                        } else {
-                            Blockchain::push(blockchain, Block::Micro(block))
-                        };
-
-                        if let Err(e) = &result {
-                            error!("Failed to push our own block onto the chain: {:?}", e);
-                        }
-
-                        let event = result
-                            .map(move |result| ProduceMicroBlockEvent::MicroBlock(block1, result))
-                            .ok();
-                        break Some(event);
-                    } else {
-                        delay = Duration::from_millis(expected_next_ts - now);
-                    };
-                } else {
-                    break None;
+            if !in_current_state(&blockchain.head()) {
+                break Some(None);
+            } else if !self.is_our_turn(&blockchain) {
+                break None;
+            }
+
+            // We want to produce a block at the expected timestamp for this block in this batch
+            // as it is calculated by the reward function and set the producer timeout accordingly
+            let now = systemtime_to_timestamp(SystemTime::now());
+
+            // If the timestamp hasn't passed, wait until the expected block timestamp
+            // to produce the block.
+            if expected_next_ts > now {
+                delay = Duration::from_millis(expected_next_ts - now);
+                continue;
+            }
+
+            // If the expected timestamp is already in the past, produce a block immediately.
+            info!(
+                block_number = self.block_number,
+                slot_band = self.validator_slot_band,
+                "Our turn, producing micro block #{}",
+                self.block_number,
+            );
+
+            let block = match self.produce_micro_block(&blockchain) {
+                Ok(block) => block,
+                Err(error) => {
+                    error!(
+                        block_number = self.block_number,
+                        %error,
+                        "Failed to construct micro block"
+                    );
+
+                    // Add an artificial panic here so that it's easier to catch any remaining
+                    // mempool bugs.
+                    // TODO Remove this.
+                    panic!("Failed to construct micro block");
+
+                    // Sleep a bit to allow the task to be cancelled externally if needed.
+                    delay = Duration::from_millis(50);
```

Comment:
Mark this to allow unreachable code, for the time that the above panic is in place.